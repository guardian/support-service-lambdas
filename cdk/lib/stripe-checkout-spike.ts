import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import {
	type GuFunctionProps,
	GuLambdaFunction,
} from '@guardian/cdk/lib/constructs/lambda';
import { type App, Duration } from 'aws-cdk-lib';
import { EventBus, Rule } from 'aws-cdk-lib/aws-events';
import { SqsQueue } from 'aws-cdk-lib/aws-events-targets';
import { Effect, PolicyStatement, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { SqsEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import { Queue } from 'aws-cdk-lib/aws-sqs';
import { nodeVersion } from './node-version';

export class StripeCheckoutSpike extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const app = 'stripe-checkout-spike';

		const stripeBus = EventBus.fromEventBusArn(
			this,
			'StripeBus',
			`arn:aws:events:${this.region}::event-source/aws.partner/stripe.com/ed_test_61RqARF6hmlRU6YCC16Rq9XTM0E9VxCGACIr7p1tg5ZY`,
		);

		const deadLetterQueue = new Queue(this, `dead-letters-${app}-queue`, {
			queueName: `dead-letters-${app}-queue-${props.stage}`,
			retentionPeriod: Duration.days(14),
		});

		const queue = new Queue(this, `${app}-queue`, {
			queueName: `${app}-queue-${props.stage}`,
			deadLetterQueue: {
				queue: deadLetterQueue,
				maxReceiveCount: 1,
			},
		});

		const rule = new Rule(this, 'StripeBusToQueueRule', {
			description: 'Send stripe events to queue',
			eventPattern: {
				region: [this.region],
				account: [this.account],
			},
			eventBus: stripeBus,
			targets: [new SqsQueue(queue)],
		});

		const sendMessagePolicyStatement = new PolicyStatement({
			sid: `Allow stripe bus to send messages to the ${app}-queue`,
			principals: [new ServicePrincipal('events.amazonaws.com')],
			effect: Effect.ALLOW,
			resources: [queue.queueArn],
			actions: ['sqs:SendMessage'],
			conditions: {
				ArnEquals: {
					'aws:SourceArn': rule.ruleArn,
				},
			},
		});

		queue.addToResourcePolicy(sendMessagePolicyStatement);

		const lambdaDefaultConfig: Pick<
			GuFunctionProps,
			'app' | 'memorySize' | 'fileName' | 'runtime' | 'timeout' | 'environment'
		> = {
			app,
			runtime: nodeVersion,
			fileName: `${app}.zip`,
			environment: {
				APP: app,
				STACK: this.stack,
				STAGE: this.stage,
				BRAZE_EMAILS_QUEUE_URL: `https://sqs.${this.region}.amazonaws.com/${this.account}/braze-emails-${this.stage}`,
			},
		};

		new GuLambdaFunction(this, 'checkoutSessionCompletedEventHandler', {
			...lambdaDefaultConfig,
			handler: 'checkoutSessionCompletedEventHandler.handler',
			functionName: `checkout-session-completed-event-handler-${this.stage}`,
			environment: { ...lambdaDefaultConfig.environment },
			events: [new SqsEventSource(queue)],
			initialPolicy: [
				new PolicyStatement({
					actions: [
						'secretsmanager:GetSecretValue',
						'secretsmanager:DescribeSecret',
					],
					resources: [
						`arn:aws:secretsmanager:${this.region}:${this.account}:secret:${this.stage}/Identity/stripe-checkout-spike-*`,
					],
				}),
				new PolicyStatement({
					actions: ['sqs:GetQueueUrl', 'sqs:SendMessage'],
					resources: [
						`arn:aws:sqs:${this.region}:${this.account}:braze-emails-${this.stage}`,
					],
				}),
			],
		});
	}
}
