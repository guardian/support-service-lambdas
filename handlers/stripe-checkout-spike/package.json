{
	"name": "stripe-checkout-spike",
	"description": "Stripe checkout spike.",
	"scripts": {
		"test": "jest --group=-integration",
		"it-test": "jest --group=integration",
		"build": "esbuild --bundle --platform=node --target=node20 --outdir=target src/handlers/*.ts",
		"lint": "eslint src/**/*.ts test/**/*.ts",
		"package": "pnpm type-check && pnpm lint && pnpm check-formatting && pnpm test && pnpm build && cd target && zip -qr stripe-checkout-spike.zip ./*.js",
		"type-check": "tsc --noEmit",
		"check-formatting": "prettier --check **.ts",
		"fix-formatting": "prettier --write **.ts"
	},
	"devDependencies": {
		"@types/aws-lambda": "^8.10.147",
		"aws-sdk-client-mock": "4.1.0"
	},
	"dependencies": {
		"@aws-sdk/client-dynamodb": "^3.734.0",
		"@aws-sdk/client-eventbridge": "^3.741.0",
		"@aws-sdk/client-secrets-manager": "3.682.0",
		"@aws-sdk/client-sqs": "3.699.0",
		"stripe": "^17.5.0"
	}
}
