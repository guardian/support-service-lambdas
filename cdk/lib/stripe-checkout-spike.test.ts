import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { StripeCheckoutSpike } from './stripe-checkout-spike';

describe('The StripeCheckoutSpike stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const codeStack = new StripeCheckoutSpike(
			app,
			`stripe-checkout-spike-CODE`,
			{
				stack: 'support',
				stage: 'CODE',
			},
		);
		const prodStack = new StripeCheckoutSpike(
			app,
			`stripe-checkout-spike-PROD`,
			{
				stack: 'support',
				stage: 'PROD',
			},
		);
		expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
		expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
	});
});
