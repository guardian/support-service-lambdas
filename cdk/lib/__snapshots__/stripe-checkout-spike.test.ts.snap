// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The StripeCheckoutSpike stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "StripeBusToQueueRule315DB460": {
      "Properties": {
        "Description": "Send stripe events to queue",
        "EventBusName": "aws.partner/stripe.com/ed_test_61RqARF6hmlRU6YCC16Rq9XTM0E9VxCGACIr7p1tg5ZY",
        "EventPattern": {
          "account": [
            {
              "Ref": "AWS::AccountId",
            },
          ],
          "region": [
            {
              "Ref": "AWS::Region",
            },
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "stripecheckoutspikequeue88DF5386",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "checkoutSessionCompletedEventHandlerF07DFDDB": {
      "DependsOn": [
        "checkoutSessionCompletedEventHandlerServiceRoleDefaultPolicy249BD397",
        "checkoutSessionCompletedEventHandlerServiceRoleC9F47EDB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "support/CODE/stripe-checkout-spike/stripe-checkout-spike.zip",
        },
        "Environment": {
          "Variables": {
            "ACQUISITION_BUS_NAME": "acquisitions-bus-CODE",
            "APP": "stripe-checkout-spike",
            "BRAZE_EMAILS_QUEUE_URL": {
              "Fn::Join": [
                "",
                [
                  "https://sqs.",
                  {
                    "Ref": "AWS::Region",
                  },
                  ".amazonaws.com/",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  "/braze-emails-CODE",
                ],
              ],
            },
            "CONTRIBUTIONS_STORE_QUEUE_URL": {
              "Fn::Join": [
                "",
                [
                  "https://sqs.",
                  {
                    "Ref": "AWS::Region",
                  },
                  ".amazonaws.com/",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  "/contributions-store-queue-CODE",
                ],
              ],
            },
            "IDENTITY_API_URL": "https://idapi.code.dev-theguardian.com",
            "SOFT_OPT_IN_CONSENT_QUEUE_URL": {
              "Fn::Join": [
                "",
                [
                  "https://sqs.",
                  {
                    "Ref": "AWS::Region",
                  },
                  ".amazonaws.com/",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  "/soft-opt-in-consent-setter-queue-CODE",
                ],
              ],
            },
            "STACK": "support",
            "STAGE": "CODE",
            "SUPPORTER_PRODUCT_DATA_TABLE_NAME": "SupporterProductData-CODE",
          },
        },
        "FunctionName": "checkout-session-completed-event-handler-CODE",
        "Handler": "checkoutSessionCompletedEventHandler.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "checkoutSessionCompletedEventHandlerServiceRoleC9F47EDB",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "stripe-checkout-spike",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "checkoutSessionCompletedEventHandlerServiceRoleC9F47EDB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "stripe-checkout-spike",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "checkoutSessionCompletedEventHandlerServiceRoleDefaultPolicy249BD397": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:secretsmanager:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":secret:CODE/Identity/stripe-checkout-spike-*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "sqs:GetQueueUrl",
                "sqs:SendMessage",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:sqs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":braze-emails-CODE",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:sqs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":contributions-store-queue-CODE",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:sqs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":soft-opt-in-consent-setter-queue-CODE",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "events:PutEvents",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:events:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":event-bus/acquisitions-bus-CODE",
                  ],
                ],
              },
            },
            {
              "Action": [
                "dynamodb:GetItem",
                "dynamodb:PutItem",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:dynamodb:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":table/SupporterProductData-CODE",
                  ],
                ],
              },
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "stripecheckoutspikequeue88DF5386",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/support/CODE/stripe-checkout-spike/stripe-checkout-spike.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/support/stripe-checkout-spike",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/support/stripe-checkout-spike/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "checkoutSessionCompletedEventHandlerServiceRoleDefaultPolicy249BD397",
        "Roles": [
          {
            "Ref": "checkoutSessionCompletedEventHandlerServiceRoleC9F47EDB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "checkoutSessionCompletedEventHandlerSqsEventSourcestripecheckoutspikeCODEstripecheckoutspikequeue7C9AD88C34BEC033": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "stripecheckoutspikequeue88DF5386",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "checkoutSessionCompletedEventHandlerF07DFDDB",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "deadlettersstripecheckoutspikequeue18B0557F": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "dead-letters-stripe-checkout-spike-queue-CODE",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "stripecheckoutspikequeue88DF5386": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "stripe-checkout-spike-queue-CODE",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "deadlettersstripecheckoutspikequeue18B0557F",
              "Arn",
            ],
          },
          "maxReceiveCount": 1,
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "stripecheckoutspikequeuePolicy24824C45": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "StripeBusToQueueRule315DB460",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "stripecheckoutspikequeue88DF5386",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "StripeBusToQueueRule315DB460",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "stripecheckoutspikequeue88DF5386",
                  "Arn",
                ],
              },
              "Sid": "Allow stripe bus to send messages to the stripe-checkout-spike-queue",
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "stripecheckoutspikequeue88DF5386",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
  },
}
`;

exports[`The StripeCheckoutSpike stack matches the snapshot 2`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "StripeBusToQueueRule315DB460": {
      "Properties": {
        "Description": "Send stripe events to queue",
        "EventBusName": "aws.partner/stripe.com/ed_test_61RqARF6hmlRU6YCC16Rq9XTM0E9VxCGACIr7p1tg5ZY",
        "EventPattern": {
          "account": [
            {
              "Ref": "AWS::AccountId",
            },
          ],
          "region": [
            {
              "Ref": "AWS::Region",
            },
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "stripecheckoutspikequeue88DF5386",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "checkoutSessionCompletedEventHandlerF07DFDDB": {
      "DependsOn": [
        "checkoutSessionCompletedEventHandlerServiceRoleDefaultPolicy249BD397",
        "checkoutSessionCompletedEventHandlerServiceRoleC9F47EDB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "support/PROD/stripe-checkout-spike/stripe-checkout-spike.zip",
        },
        "Environment": {
          "Variables": {
            "ACQUISITION_BUS_NAME": "acquisitions-bus-PROD",
            "APP": "stripe-checkout-spike",
            "BRAZE_EMAILS_QUEUE_URL": {
              "Fn::Join": [
                "",
                [
                  "https://sqs.",
                  {
                    "Ref": "AWS::Region",
                  },
                  ".amazonaws.com/",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  "/braze-emails-PROD",
                ],
              ],
            },
            "CONTRIBUTIONS_STORE_QUEUE_URL": {
              "Fn::Join": [
                "",
                [
                  "https://sqs.",
                  {
                    "Ref": "AWS::Region",
                  },
                  ".amazonaws.com/",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  "/contributions-store-queue-PROD",
                ],
              ],
            },
            "IDENTITY_API_URL": "https://idapi.theguardian.com",
            "SOFT_OPT_IN_CONSENT_QUEUE_URL": {
              "Fn::Join": [
                "",
                [
                  "https://sqs.",
                  {
                    "Ref": "AWS::Region",
                  },
                  ".amazonaws.com/",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  "/soft-opt-in-consent-setter-queue-PROD",
                ],
              ],
            },
            "STACK": "support",
            "STAGE": "PROD",
            "SUPPORTER_PRODUCT_DATA_TABLE_NAME": "SupporterProductData-PROD",
          },
        },
        "FunctionName": "checkout-session-completed-event-handler-PROD",
        "Handler": "checkoutSessionCompletedEventHandler.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "checkoutSessionCompletedEventHandlerServiceRoleC9F47EDB",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "stripe-checkout-spike",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "checkoutSessionCompletedEventHandlerServiceRoleC9F47EDB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "stripe-checkout-spike",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "checkoutSessionCompletedEventHandlerServiceRoleDefaultPolicy249BD397": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:secretsmanager:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":secret:PROD/Identity/stripe-checkout-spike-*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "sqs:GetQueueUrl",
                "sqs:SendMessage",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:sqs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":braze-emails-PROD",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:sqs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":contributions-store-queue-PROD",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:sqs:",
                      {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":soft-opt-in-consent-setter-queue-PROD",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "events:PutEvents",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:events:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":event-bus/acquisitions-bus-PROD",
                  ],
                ],
              },
            },
            {
              "Action": [
                "dynamodb:GetItem",
                "dynamodb:PutItem",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:dynamodb:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":table/SupporterProductData-PROD",
                  ],
                ],
              },
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "stripecheckoutspikequeue88DF5386",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/support/PROD/stripe-checkout-spike/stripe-checkout-spike.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/support/stripe-checkout-spike",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/support/stripe-checkout-spike/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "checkoutSessionCompletedEventHandlerServiceRoleDefaultPolicy249BD397",
        "Roles": [
          {
            "Ref": "checkoutSessionCompletedEventHandlerServiceRoleC9F47EDB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "checkoutSessionCompletedEventHandlerSqsEventSourcestripecheckoutspikePRODstripecheckoutspikequeueC1B16F66366C2BB9": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "stripecheckoutspikequeue88DF5386",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "checkoutSessionCompletedEventHandlerF07DFDDB",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "deadlettersstripecheckoutspikequeue18B0557F": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "dead-letters-stripe-checkout-spike-queue-PROD",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "stripecheckoutspikequeue88DF5386": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "QueueName": "stripe-checkout-spike-queue-PROD",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "deadlettersstripecheckoutspikequeue18B0557F",
              "Arn",
            ],
          },
          "maxReceiveCount": 1,
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "support",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "stripecheckoutspikequeuePolicy24824C45": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "StripeBusToQueueRule315DB460",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "stripecheckoutspikequeue88DF5386",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "StripeBusToQueueRule315DB460",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "stripecheckoutspikequeue88DF5386",
                  "Arn",
                ],
              },
              "Sid": "Allow stripe bus to send messages to the stripe-checkout-spike-queue",
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "stripecheckoutspikequeue88DF5386",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
  },
}
`;
